File: sdk/python/examples/test_advanced_features.py
#!/usr/bin/env python3
"""
A3S Code — Advanced Features Demo

Demonstrates: auto-compact, memory, security, hooks, planning,
permission policy, and resilience options.

Usage:
    cd crates/code/sdk/python
    python examples/test_advanced_features.py

Requires a valid LLM API key in ~/.a3s/config.hcl or $A3S_CONFIG.
"""

import asyncio
import os
import tempfile
import shutil
from pathlib import Path

import a3s_code


class AdvancedFeaturesTest:
    """Tests for advanced features: auto-compact, memory, security, hooks, etc."""

    def __init__(self, config_path: str) -> None:
        self.config_path = config_path

    @staticmethod
    def find_config_path() -> str:
        if config := os.environ.get("A3S_CONFIG"):
            return config
        home_config = Path.home() / ".a3s" / "config.hcl"
        if home_config.exists():
            return str(home_config)
        raise RuntimeError("No config found. Set A3S_CONFIG or create ~/.a3s/config.hcl")

    # ========================================================================
    # Test 1: Auto-Compact
    # ========================================================================

    async def test_auto_compact(self) -> None:
        print("\n=== Test 1: Auto-Compact ===")
        agent = await a3s_code.Agent.create(self.config_path)

        opts = a3s_code.SessionOptions()
        opts.auto_compact = True
        opts.auto_compact_threshold = 0.8

        session = agent.session(tempfile.gettempdir(), options=opts)
        result = await session.send("Say hello in one sentence.")
        print(f"  auto-compact session created, response: {result.text[:60]}")

    # ========================================================================
    # Test 2: File Memory
    # ========================================================================

    async def test_memory(self) -> None:
        print("\n=== Test 2: File Memory ===")
        agent = await a3s_code.Agent.create(self.config_path)
        mem_dir = tempfile.mkdtemp(prefix="a3s-mem-")

        try:
            opts = a3s_code.SessionOptions()
            opts.memory_dir = mem_dir

            session = agent.session(tempfile.gettempdir(), options=opts)
            result = await session.send("Remember: my favorite language is Rust.")
            print(f"  memory session created, response: {result.text[:60]}")
        finally:
            shutil.rmtree(mem_dir, ignore_errors=True)

    # ========================================================================
    # Test 3: Default Security Provider
    # ========================================================================

    async def test_security(self) -> None:
        print("\n=== Test 3: Security Provider ===")
        agent = await a3s_code.Agent.create(self.config_path)

        opts = a3s_code.SessionOptions()
        opts.default_security = True

        session = agent.session(tempfile.gettempdir(), options=opts)
        result = await session.send("What is 2 + 2?")
        print(f"  security session created, response: {result.text[:60]}")

    # ========================================================================
    # Test 4: Hooks — intercept lifecycle events
    # ========================================================================

    async def test_hooks(self) -> None:
        print("\n=== Test 4: Hooks ===")
        agent = await a3s_code.Agent.create(self.config_path)
        session = agent.session(tempfile.gettempdir())

        tool_calls: list[str] = []

        def on_pre_tool_use(event: dict) -> None:
            tool_calls.append(event.get("tool_name", "unknown"))
            print(f"  -> pre_tool_use: {str(event)[:80]}")
            return None  # allow execution

        session.register_hook("log-tools", "pre_tool_use", on_pre_tool_use)
        print(f"  hook registered, hook count: {session.hook_count()}")

        session.unregister_hook("log-tools")
        print(f"  hook unregistered, hook count: {session.hook_count()}")

    # ========================================================================
    # Test 5: Planning Mode
    # ========================================================================

    async def test_planning(self) -> None:
        print("\n=== Test 5: Planning Mode ===")
        agent = await a3s_code.Agent.create(self.config_path)

        session = agent.session(
            tempfile.gettempdir(),
            planning=True,
            goal_tracking=True,
        )
        result = await session.send("List 3 benefits of Rust in one sentence each.")
        print(f"  planning session, response length: {len(result.text)}")

    # ========================================================================
    # Test 6: Permissive Policy
    # ========================================================================

    async def test_permissive(self) -> None:
        print("\n=== Test 6: Permissive Policy ===")
        agent = await a3s_code.Agent.create(self.config_path)

        session = agent.session(tempfile.gettempdir(), permissive=True)
        result = await session.send("What is the current directory?")
        print(f"  permissive session, response: {result.text[:60]}")

    # ========================================================================
    # Test 7: Resilience Options
    # ========================================================================

    async def test_resilience(self) -> None:
        print("\n=== Test 7: Resilience Options ===")
        agent = await a3s_code.Agent.create(self.config_path)

        session = agent.session(
            tempfile.gettempdir(),
            max_parse_retries=3,
            tool_timeout_ms=30000,
            circuit_breaker_threshold=5,
        )
        result = await session.send('Say "resilient" once.')
        print(f"  resilience session, response: {result.text[:60]}")

    # ========================================================================
    # Test 8: Combined — security + auto-compact + memory + hooks
    # ========================================================================

    async def test_combined(self) -> None:
        print("\n=== Test 8: Combined Features ===")
        agent = await a3s_code.Agent.create(self.config_path)
        mem_dir = tempfile.mkdtemp(prefix="a3s-combined-")

        try:
            opts = a3s_code.SessionOptions()
            opts.default_security = True
            opts.auto_compact = True
            opts.auto_compact_threshold = 0.9
            opts.memory_dir = mem_dir

            session = agent.session(tempfile.gettempdir(), options=opts, permissive=True)

            def on_post_tool_use(event: dict) -> None:
                print("  -> post_tool_use fired")
                return None

            session.register_hook("audit", "post_tool_use", on_post_tool_use)

            result = await session.send("What is 1 + 1?")
            print(f"  combined session, response: {result.text[:60]}")
        finally:
            shutil.rmtree(mem_dir, ignore_errors=True)

    # ========================================================================
    # Run All Tests
    # ========================================================================

    async def run_all(self) -> int:
        print("A3S Code -- Advanced Features Test")
        print("===================================")

        tests = [
            self.test_auto_compact,
            self.test_memory,
            self.test_security,
            self.test_hooks,
            self.test_planning,
            self.test_permissive,
            self.test_resilience,
            self.test_combined,
        ]

        passed = 0
        failed = 0

        for test in tests:
            try:
                await test()
                passed += 1
            except Exception as e:
                print(f"  {test.__name__} failed: {e}")
                failed += 1

        print(f"\n===================================")
        print(f"Results: {passed} passed, {failed} failed")
        return failed


if __name__ == "__main__":
    config_path = AdvancedFeaturesTest.find_config_path()
    suite = AdvancedFeaturesTest(config_path)
    failed = asyncio.run(suite.run_all())
    raise SystemExit(failed)

File: sdk/node/examples/test_advanced_features.ts
#!/usr/bin/env npx tsx
/**
 * A3S Code — Advanced Features Demo
 *
 * Demonstrates: auto-compact, memory, security, hooks, batch tool,
 * permission checker, planning, and context providers.
 *
 * ## Usage
 *
 *     cd crates/code/sdk/node
 *     npx tsx examples/test_advanced_features.ts
 *
 * Requires a valid LLM API key in ~/.a3s/config.hcl or $A3S_CONFIG.
 */

import { Agent, Session, AgentResult } from '../index.js';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

// ============================================================================
// AdvancedFeaturesTest
// ============================================================================

class AdvancedFeaturesTest {
  private readonly configPath: string;

  constructor(configPath: string) {
    this.configPath = configPath;
  }

  // --------------------------------------------------------------------------
  // Static helpers
  // --------------------------------------------------------------------------

  static findConfigPath(): string {
    if (process.env.A3S_CONFIG) return process.env.A3S_CONFIG;
    const homeConfig: string = path.join(os.homedir(), '.a3s', 'config.hcl');
    if (fs.existsSync(homeConfig)) return homeConfig;
    throw new Error('No config found. Set A3S_CONFIG or create ~/.a3s/config.hcl');
  }

  // --------------------------------------------------------------------------
  // Test 1: Auto-Compact
  // --------------------------------------------------------------------------

  async testAutoCompact(): Promise<void> {
    console.log('\n=== Test 1: Auto-Compact ===');
    const agent: Agent = await Agent.create(this.configPath);

    // Enable auto-compact at 80% context usage
    const session: Session = agent.session(os.tmpdir(), {
      autoCompact: true,
      autoCompactThreshold: 0.8,
    });

    const result: AgentResult = await session.send('Say hello in one sentence.');
    console.log('auto-compact session created, response:', result.text.slice(0, 60));
  }

  // --------------------------------------------------------------------------
  // Test 2: File Memory
  // --------------------------------------------------------------------------

  async testMemory(): Promise<void> {
    console.log('\n=== Test 2: File Memory ===');
    const agent: Agent = await Agent.create(this.configPath);
    const memDir: string = fs.mkdtempSync(path.join(os.tmpdir(), 'a3s-mem-'));

    const session: Session = agent.session(os.tmpdir(), {
      memoryDir: memDir,
    });

    const result: AgentResult = await session.send('Remember: my favorite language is Rust.');
    console.log('memory session created, response:', result.text.slice(0, 60));

    // Cleanup
    fs.rmSync(memDir, { recursive: true, force: true });
  }

  // --------------------------------------------------------------------------
  // Test 3: Default Security Provider
  // --------------------------------------------------------------------------

  async testSecurity(): Promise<void> {
    console.log('\n=== Test 3: Security Provider ===');
    const agent: Agent = await Agent.create(this.configPath);

    // Enable default security: taint-tracks input, sanitizes output
    const session: Session = agent.session(os.tmpdir(), {
      defaultSecurity: true,
    });

    const result: AgentResult = await session.send('What is 2 + 2?');
    console.log('security session created, response:', result.text.slice(0, 60));
  }

  // --------------------------------------------------------------------------
  // Test 4: Hooks — intercept lifecycle events
  // --------------------------------------------------------------------------

  async testHooks(): Promise<void> {
    console.log('\n=== Test 4: Hooks ===');
    const agent: Agent = await Agent.create(this.configPath);
    const session: Session = agent.session(os.tmpdir());

    const toolCalls: string[] = [];

    // Register a hook that fires before every tool use
    session.registerHook('log-tools', 'pre_tool_use', (event: Record<string, unknown>) => {
      toolCalls.push((event.toolName as string) || (event.tool_name as string) || 'unknown');
      console.log(`  -> pre_tool_use: ${JSON.stringify(event).slice(0, 80)}`);
      return null; // allow execution
    });

    console.log(`hook registered, hook count: ${session.hookCount()}`);

    // Cleanup
    session.unregisterHook('log-tools');
    console.log(`hook unregistered, hook count: ${session.hookCount()}`);
  }

  // --------------------------------------------------------------------------
  // Test 5: Planning Mode
  // --------------------------------------------------------------------------

  async testPlanning(): Promise<void> {
    console.log('\n=== Test 5: Planning Mode ===');
    const agent: Agent = await Agent.create(this.configPath);

    const session: Session = agent.session(os.tmpdir(), {
      planning: true,
      goalTracking: true,
    });

    const result: AgentResult = await session.send('List 3 benefits of Rust in one sentence each.');
    console.log('planning session, response length:', result.text.length);
  }

  // --------------------------------------------------------------------------
  // Test 6: Permissive Policy (no HITL confirmation)
  // --------------------------------------------------------------------------

  async testPermissive(): Promise<void> {
    console.log('\n=== Test 6: Permissive Policy ===');
    const agent: Agent = await Agent.create(this.configPath);

    const session: Session = agent.session(os.tmpdir(), {
      permissive: true,
    });

    const result: AgentResult = await session.send('What is the current directory?');
    console.log('permissive session, response:', result.text.slice(0, 60));
  }

  // --------------------------------------------------------------------------
  // Test 7: Resilience Options
  // --------------------------------------------------------------------------

  async testResilience(): Promise<void> {
    console.log('\n=== Test 7: Resilience Options ===');
    const agent: Agent = await Agent.create(this.configPath);

    const session: Session = agent.session(os.tmpdir(), {
      maxParseRetries: 3,
      toolTimeoutMs: 30000,
      circuitBreakerThreshold: 5,
    });

    const result: AgentResult = await session.send('Say "resilient" once.');
    console.log('resilience session, response:', result.text.slice(0, 60));
  }

  // --------------------------------------------------------------------------
  // Test 8: Combined — security + auto-compact + memory + hooks
  // --------------------------------------------------------------------------

  async testCombined(): Promise<void> {
    console.log('\n=== Test 8: Combined Features ===');
    const agent: Agent = await Agent.create(this.configPath);
    const memDir: string = fs.mkdtempSync(path.join(os.tmpdir(), 'a3s-combined-'));

    const session: Session = agent.session(os.tmpdir(), {
      defaultSecurity: true,
      autoCompact: true,
      autoCompactThreshold: 0.9,
      memoryDir: memDir,
      permissive: true,
    });

    session.registerHook('audit', 'post_tool_use', (event: Record<string, unknown>) => {
      console.log(`  -> post_tool_use fired`);
      return null;
    });

    const result: AgentResult = await session.send('What is 1 + 1?');
    console.log('combined session, response:', result.text.slice(0, 60));

    fs.rmSync(memDir, { recursive: true, force: true });
  }

  // --------------------------------------------------------------------------
  // Run All
  // --------------------------------------------------------------------------

  async runAll(): Promise<void> {
    const tests: Array<{ name: string; fn: () => Promise<void> }> = [
      { name: 'testAutoCompact', fn: () => this.testAutoCompact() },
      { name: 'testMemory', fn: () => this.testMemory() },
      { name: 'testSecurity', fn: () => this.testSecurity() },
      { name: 'testHooks', fn: () => this.testHooks() },
      { name: 'testPlanning', fn: () => this.testPlanning() },
      { name: 'testPermissive', fn: () => this.testPermissive() },
      { name: 'testResilience', fn: () => this.testResilience() },
      { name: 'testCombined', fn: () => this.testCombined() },
    ];

    let passed: number = 0;
    let failed: number = 0;

    for (const test of tests) {
      try {
        await test.fn();
        passed++;
      } catch (err) {
        const message: string = err instanceof Error ? err.message : String(err);
        console.error(`  ${test.name} failed:`, message);
        failed++;
      }
    }

    console.log(`\n===================================`);
    console.log(`Results: ${passed} passed, ${failed} failed`);
    process.exit(failed > 0 ? 1 : 0);
  }
}

// ============================================================================
// Main
// ============================================================================

async function main(): Promise<void> {
  console.log('A3S Code — Advanced Features Test');
  console.log('===================================');

  const configPath: string = AdvancedFeaturesTest.findConfigPath();
  const test = new AdvancedFeaturesTest(configPath);
  await test.runAll();
}

main().catch((err: unknown) => {
  console.error('Fatal:', err);
  process.exit(1);
});
